feature: support torch_distributed distribution for Trainium instances (#3424)
aws/sagemaker-python-sdk: doc/frameworks/pytorch/using_pytorch.rst

    pt_estimator.fit("s3://bucket/path/to/training/data")

.. _distributed-pytorch-training-on-trainium:

> =============================================

Reviewer: Need to bring the title back. `.. _distributed-pytorch-training-on-trainium:` is only a flag that enables links.

> =============================================

Reviewer: ```suggestion

Distributed PyTorch Training on Trainium
========================================
```

Author: Reverted this change to earlier version. Can you check now?

    pt_estimator.fit("s3://bucket/path/to/training/data")

> Distributed PyTorch Training on Trainum (trn1) Instances

Reviewer: ```suggestion
.. _distributed-pytorch-training-on-trainium:

Distributed PyTorch Training on Trainum (trn1) Instances
```

Reviewer: This is to make sure the link you added in `estimator.py` works.

========================================================

SageMaker Training on Trainium instances now supports the `xla`
package through `torchrun`. With this, you do not need to manually pass RANK, 
WORLD_SIZE, MASTER_ADDR, and MASTER_PORT. You can launch the training job using the
:class:`sagemaker.pytorch.estimator.PyTorch` estimator class
with the ``torch_distributed`` option as the distribution strategy.

.. note::

>   This torch_distributed support is available

Reviewer: ```suggestion
  This ``torch_distributed`` support is available
```

>   in the SageMaker Trainium (trn1) PyTorch Deep Learning Containers starting v1.11.0

Reviewer: ```suggestion
  in the SageMaker Trainium (trn1) PyTorch Deep Learning Containers starting v1.11.0.
```

>   `Supported Containers <https://github.com/aws/deep-learning-containers/blob/master/available_images.md#neuron-containers>`_

Reviewer: ```suggestion
  To find a complete list of supported versions of PyTorch Neuron, see `Neuron Containers <https://github.com/aws/deep-learning-containers/blob/master/available_images.md#neuron-containers>`_ in the *AWS Deep Learning Containers GitHub repository*.
```

  SageMaker Debugger and Profiler are currently not supported with Trainium instances.

Adapt Your Training Script
--------------------------

To initialize distributed training in your script, call
`torch.distributed.init_process_group
<https://pytorch.org/docs/master/distributed.html#torch.distributed.init_process_group>`_
with the ``xla`` backend as shown below.

.. code:: python

    import torch.distributed as dist

    dist.init_process_group('xla')

> SageMaker takes care of ``'MASTER_ADDR'`` and ``'MASTER_PORT'`` for you via `torchrun`

Reviewer: ```suggestion
SageMaker takes care of ``'MASTER_ADDR'`` and ``'MASTER_PORT'`` for you via ``torchrun``.
```

> For detailed documentation about modifying your training script for Trainium. Please refer `AWS Neuron Documentation <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/frameworks/torch/torch-neuronx/tutorials/training/mlp.html?highlight=torchrun#id5>`_

Reviewer: ```suggestion
For detailed documentation about modifying your training script for Trainium, see `Multi-worker data-parallel MLP training using torchrun <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/frameworks/torch/torch-neuronx/tutorials/training/mlp.html?highlight=torchrun#multi-worker-data-parallel-mlp-training-using-torchrun>`_ in the *AWS Neuron Documentation*.
```

**Currently Supported backends:**

-  ``xla`` for Trainium (Trn1) instances

> For up-to-date information on supported backends for Trainium instances, please refer `AWS Neuron Documentation <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/index.html>`_

Reviewer: ```suggestion
For up-to-date information on supported backends for Trainium instances, see `AWS Neuron Documentation <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/index.html>`_.
```

Launching a Distributed Training Job
------------------------------------

You can run multi-node distributed PyTorch training jobs on Trainium instances using the
:class:`sagemaker.pytorch.estimator.PyTorch` estimator class.
With ``instance_count=1``, the estimator submits a
single-node training job to SageMaker; with ``instance_count`` greater
than one, a multi-node training job is launched.

With the ``torch_distributed`` option, the SageMaker PyTorch estimator runs a SageMaker

> training container for PyTorch, sets up the environment, and launches

Reviewer: ```suggestion
training container for PyTorch Neuron, sets up the environment, and launches
```

the training job using the ``torchrun`` command on each worker with the given information
during the PyTorch DDP initialization.

.. note::

> The following example shows how to run a PyTorch training using `torch_distributed` in SageMaker

Reviewer: ```suggestion
The following example shows how to run a PyTorch training using ``torch_distributed`` in SageMaker
```
